File: Ruby/023.rb
#!/usr/bin/env ruby
# encoding: utf-8
require "set"
require_relative "euler"

limit = 20162
ans = 0
abn_nums = Set.new

(1 ... limit).each do |num|
  abn_nums << num if num.proper_divisor_sum > num

  flag = true
  abn_nums.each do |abn|
    if abn_nums.include?(num - abn)
      flag = false
      break
    end
  end

  ans += num if flag
end

puts ans

File: Python/023.py
#!/usr/bin/env python
# encoding: utf-8
from euler import proper_divisor_sum


def main():
    limit, result = 20162, 0
    abundant_numbers = set()

    for num in range(1, limit):
        if proper_divisor_sum(num) > num:
            abundant_numbers.add(num)

        if not any((num - x) in abundant_numbers for x in abundant_numbers):
            result += num
    print(result)


if __name__ == '__main__':
    main()
